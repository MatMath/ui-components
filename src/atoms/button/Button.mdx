---
name: Button
menu: Components
---

import { Playground, Props } from 'docz';
import Button from './Button';

# Button

A Button that allows users to take actions, and make choices, with a single tap. It can be a either:

- `Contained (filled) button`
- `Text button`

#### Downside: Since Button props extends the default html button props, Docz will display all props including `React.ButtonHTMLAttributes` ones ðŸ˜­

Bug on repo link: https://github.com/doczjs/docz/issues/895

## Properties

<Props of={Button} />

## Usage

### Default variant for the button is `contained`

<Playground style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-around' }}>
  <div>
    <Button nature="danger">Danger Button</Button>
  </div>

<div>
  <Button nature='dark'>Dark Button</Button>
</div>

<div>
  <Button nature='light'>Light Button</Button>
</div>

<div>
  <Button nature='secondary'>Secondary Button</Button>
</div>

  <div>
    <Button nature="success">Success Button</Button>
  </div>
</Playground>

### You can have a text button by specifying `variant` property to `text`

<Playground style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-around' }}>
  <div>
    <Button nature="danger" variant="text">Danger Button</Button>
  </div>

<div>
  <Button nature='dark' variant='text'>
    Dark Button
  </Button>
</div>

<div>
  <Button nature='light' variant='text'>
    Light Button
  </Button>
</div>

<div>
  <Button nature='secondary' variant='text'>
    Secondary Button
  </Button>
</div>

  <div>
    <Button nature="success" variant="text">Success Button</Button>
  </div>
</Playground>

### Buttons (either `contained` or `text`) can be `large` ones

<Playground style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-around' }}>
<div>
  <Button nature='secondary' variant='text' large={true}>
    Label
  </Button>
</div>

  <div>
    <Button nature="success" large={true}>Label</Button>
  </div>
</Playground>

### In disabled mode, buttons will behave as following

<Playground className="test" style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-around', flexFlow: 'wrap' }}>
<div style={{ display: 'flex', justifyContent: 'space-around', width: '100%' }}>
  <div>
    <Button nature="danger" disabled>Danger Button</Button>
  </div>

<div>
  <Button nature='dark' disabled>
    Dark Button
  </Button>
</div>

<div>
  <Button nature='light' disabled>
    Light Button
  </Button>
</div>

<div>
  <Button nature='secondary' disabled>
    Secondary Button
  </Button>
</div>

  <div>
    <Button nature="success" disabled>Success Button</Button>
  </div>
</div>

<div style={{ display: 'flex', justifyContent: 'space-around', width: '100%', margin: '10px 0' }}>
  <div>
    <Button nature="danger" variant="text" disabled>Danger Button</Button>
  </div>

<div>
  <Button nature='dark' variant='text' disabled>
    Dark Button
  </Button>
</div>

<div>
  <Button nature='light' variant='text' disabled>
    Light Button
  </Button>
</div>

<div>
  <Button nature='secondary' variant='text' disabled>
    Secondary Button
  </Button>
</div>

<div>
  <Button nature="success" variant="text" disabled>Success Button</Button>
  </div>
</div>
</Playground>
