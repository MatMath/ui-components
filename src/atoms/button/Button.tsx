import * as React from 'react';
import {
  ButtonDanger,
  ButtonDark,
  ButtonLight,
  ButtonSecondary,
  ButtonSuccess
} from '@ions';
import { assertUnreachable } from '@utility/helpers';

// Downside: Docz will display all props including React.ButtonHTMLAttributes ones
// https://github.com/doczjs/docz/issues/895
export interface BaseButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * Default is "false"
   */
  disabled?: boolean;
  /**
   * Default is "contained"
   */
  variant?: ButtonVariant;
  /**
   * Default is "false"
   */
  large?: boolean;
  children?: React.ReactNode;
}

interface Props extends BaseButtonProps {
  nature: ButtonTypes;
}

export const enum ButtonVariant {
  contained = 'contained',
  text = 'text'
}

export const enum ButtonTypes {
  DANGER = 'danger',
  SECONDARY = 'secondary',
  DARK = 'dark',
  LIGHT = 'light',
  SUCCESS = 'success'
}

const Button: React.FunctionComponent<Props> = (props: Props): JSX.Element => {
  const renderButtonByType = ({ nature, ...props }: Props): JSX.Element => {
    switch (nature) {
      case ButtonTypes.DANGER:
        return <ButtonDanger {...props} />;
      case ButtonTypes.SECONDARY:
        return <ButtonSecondary {...props} />;
      case ButtonTypes.DARK:
        return <ButtonDark {...props} />;
      case ButtonTypes.LIGHT:
        return <ButtonLight {...props} />;
      case ButtonTypes.SUCCESS:
        return <ButtonSuccess {...props} />;
      default:
        return assertUnreachable(nature);
    }
  };

  return renderButtonByType(props);
};

export default Button;
