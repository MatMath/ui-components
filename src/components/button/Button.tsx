import * as React from 'react';
import { getClassNames } from '@utility/cssUtils';
import styles from './Button.module.scss';

type ButtonNature = 'primary' | 'secondary' | 'danger' | 'ghost';
type ButtonSize = 'small' | 'medium' | 'large';
type ButtonTheme = 'light' | 'dark';

const defaultNature: ButtonNature = 'primary';
const defaultSize: ButtonSize = 'medium';
const defaultTheme: ButtonTheme = 'light';

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  nature?: ButtonNature;
  size?: ButtonSize;
  theme?: ButtonTheme;
}

const Button: React.FC<ButtonProps> = ({
  nature = defaultNature,
  size = defaultSize,
  theme = defaultTheme,
  children,
  ...htmlButtonProps
}) => {
  return (
    <button
      {...htmlButtonProps}
      className={getClassNames(
        styles.button,
        styles[nature],
        styles[size],
        styles[theme]
      )}
    >
      {children ? <span className={styles.content}>{children}</span> : null}
    </button>
  );
};

Button.defaultProps = {
  nature: defaultNature,
  size: defaultSize,
  theme: defaultTheme
};

export default Button;
