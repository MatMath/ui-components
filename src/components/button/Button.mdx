---
name: Button
menu: Components
---

import { Playground, Props } from 'docz';
import { Button } from './Button';
import { CopyPasswordIcon } from '../../atoms/icons/CopyPasswordIcon';
import { white, midGreen00, dashGreen00 } from '../../colors/colors.ts';
import '../../utility/playgroundStyles.css';

# Button

A Button that allows users to take actions, and make choices, with a single tap.

## Properties

<Props of={Button} />

## Usage

### Nature

<Playground className={{preview: 'playground-base'}}>

    <Button nature='primary'>Primary</Button>
    <Button nature='secondary'>Secondary</Button>
    <Button nature='danger'>Danger</Button>
    <Button nature='ghost'>Ghost</Button>

</Playground>

### Possibilities

All buttons can display icons, text or mixed content.

<Playground className={{preview: 'playground-base'}}>
    <Button nature='primary'>
      <CopyPasswordIcon color={white} style={{ marginRight: '8px'}}/>
      Save
    </Button>

    <Button nature='secondary'>
      <CopyPasswordIcon color={midGreen00} style={{ marginRight: '8px'}}/>
      Cancel
    </Button>

    <Button nature='danger'>
      <CopyPasswordIcon color={white} style={{ marginRight: '8px'}}/>
      Delete
    </Button>

    <Button nature='ghost'>Decline</Button>

    <Button nature='ghost'>
      <CopyPasswordIcon color={midGreen00} />
    </Button>

</Playground>

### Theme

The default theme is `light` (white), but some buttons can be displayed in a `dark` theme (Dashlane's dash green, not an actual _dark_ theme).

<Playground className={{preview: 'playground-base'}} style={{ preview: {backgroundColor: dashGreen00} }}>
    <Button nature='primary' theme='dark'>
      Save
    </Button>

    <Button nature='secondary' theme='dark'>
      Cancel
    </Button>

    <Button nature='ghost' theme='dark'>
      Decline
    </Button>

    <Button nature='ghost' size='small' theme='dark'>
      <CopyPasswordIcon color={white} />
    </Button>

</Playground>

> ⚠️ `danger` nature is not supported in `dark` theme.

### Size

Buttons can be have different sizes [`small`, `medium`, `large`] with `medium` being the default value if none is specified.

<Playground className={{preview: 'playground-base'}} style={{ preview: {minHeight: '150px'} }}>
    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between' }}>
      <Button nature='primary' size='small'>Save</Button>
      <Button nature='primary' size='medium'>Save</Button>
      <Button nature='primary' size='large'>Save</Button>
    </div>

    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between' }}>
      <Button nature='secondary' size='small'>Cancel</Button>
      <Button nature='secondary' size='medium'>Cancel</Button>
      <Button nature='secondary' size='large'>Cancel</Button>
    </div>

    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between' }}>
      <Button nature='danger' size='small'>Delete</Button>
      <Button nature='danger' size='medium'>Delete</Button>
      <Button nature='danger' size='large'>Delete</Button>
    </div>

    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'space-between' }}>
      <Button nature='ghost' size='small'>Decline</Button>
      <Button nature='ghost' size='medium'>Decline</Button>
      <Button nature='ghost' size='large'>Decline</Button>
    </div>

</Playground>

### Disabled

In disabled pseudo-state, buttons behave as following.

<Playground className={{preview: 'playground-base'}}>

    <Button nature='primary' disabled>Save</Button>
    <Button nature='secondary' disabled>Cancel</Button>
    <Button nature='danger' disabled>Delete</Button>
    <Button nature='ghost' size='medium' disabled>Decline</Button>
    <Button nature='ghost' disabled>
      <CopyPasswordIcon color={midGreen00}/>
    </Button>

</Playground>
