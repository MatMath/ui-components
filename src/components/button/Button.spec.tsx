import * as React from 'react';
import { mount } from 'enzyme';
import 'jest-styled-components';
import { Button } from './Button';

describe('<Button>', () => {
  describe('Global render', () => {
    it('should render default', () => {
      const button = mount(<Button />);

      expect(button).toMatchSnapshot();
    });
  });

  describe('Style properties', () => {
    it('should render small size', () => {
      const button = mount(<Button size='small' />);

      expect(button).toMatchSnapshot();
    });

    it('should render medium size', () => {
      const button = mount(<Button size='medium' />);

      expect(button).toMatchSnapshot();
    });

    it('should render large size', () => {
      const button = mount(<Button size='large' />);

      expect(button).toMatchSnapshot();
    });

    it('should render primary nature', () => {
      const button = mount(<Button nature='primary' />);

      expect(button).toMatchSnapshot();
    });

    it('should render secondary nature', () => {
      const button = mount(<Button nature='secondary' />);

      expect(button).toMatchSnapshot();
    });

    it('should render danger nature', () => {
      const button = mount(<Button nature='danger' />);

      expect(button).toMatchSnapshot();
    });

    it('should render danger nature and small size', () => {
      const button = mount(<Button nature='danger' size='small' />);

      expect(button).toMatchSnapshot();
    });
  });

  describe('Children render', () => {
    it('should render Default when no children are specified', () => {
      const button = mount(<Button />);

      expect(button.text()).toBe('');
    });

    it('should render the specified text', () => {
      const text = 'Hello There';
      const button = mount(<Button>{text}</Button>);

      expect(button.text()).toEqual(text);
      expect(button.find('span')).toBeTruthy();
    });

    it('should render the specified children', () => {
      const text = 'Ok';
      const button = mount(
        <Button>
          <label>{text}</label>
        </Button>
      );

      expect(button.find('label')).toBeTruthy();
    });
  });

  describe('Interaction', () => {
    it('should execute action on click', () => {
      const button = mount(<Button onClick={() => window.alert('hello')} />);

      window.alert = jest.fn();
      button.simulate('click');
      expect(window.alert).toHaveBeenCalledTimes(1);
      expect(window.alert).toHaveBeenCalledWith('hello');
    });

    it('should not execute action on click when the button is disabled', () => {
      const button = mount(
        <Button onClick={() => window.alert('hello')} disabled />
      );

      window.alert = jest.fn();
      button.simulate('click');
      expect(window.alert).toHaveBeenCalledTimes(0);
    });
  });
});
