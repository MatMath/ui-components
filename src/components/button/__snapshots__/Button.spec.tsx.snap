// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> Children render should render the specified children 1`] = `
<DocumentFragment>
  <button
    class="button primary medium light"
  >
    <span
      class="content"
    >
      <label>
        Ok
      </label>
    </span>
  </button>
</DocumentFragment>
`;

exports[`<Button> Children render should render the specified text 1`] = `
<DocumentFragment>
  <button
    class="button primary medium light"
  >
    <span
      class="content"
    >
      Hello There
    </span>
  </button>
</DocumentFragment>
`;

exports[`<Button> Global render should render default 1`] = `
<DocumentFragment>
  <button
    class="button primary medium light"
  />
</DocumentFragment>
`;

exports[`<Button> Style properties should render danger nature 1`] = `
<DocumentFragment>
  <button
    class="button danger medium light"
  />
</DocumentFragment>
`;

exports[`<Button> Style properties should render danger nature and small size 1`] = `
<DocumentFragment>
  <button
    class="button danger small light"
  />
</DocumentFragment>
`;

exports[`<Button> Style properties should render large size 1`] = `
<DocumentFragment>
  <button
    class="button primary large light"
  />
</DocumentFragment>
`;

exports[`<Button> Style properties should render medium size 1`] = `
<DocumentFragment>
  <button
    class="button primary medium light"
  />
</DocumentFragment>
`;

exports[`<Button> Style properties should render primary nature 1`] = `
<DocumentFragment>
  <button
    class="button primary medium light"
  />
</DocumentFragment>
`;

exports[`<Button> Style properties should render secondary nature 1`] = `
<DocumentFragment>
  <button
    class="button secondary medium light"
  />
</DocumentFragment>
`;

exports[`<Button> Style properties should render small size 1`] = `
<DocumentFragment>
  <button
    class="button primary medium light"
  />
</DocumentFragment>
`;
